docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,478 @@
+[Open Enclave Init-time Configuration Interface](InitTimeConfigurationInterface.md)
+to support init-time custom claims.
+The core idea of the design is to avoid breaking changes to the attestation plugin
+framework. More specifically, the `oe_get_attestation_certificate_with_evidence_v3`
+API appends the `inittime_custom_claims_buffer` to the evidence buffer, which is the output
+of the `oe_get_evidence` API. After that, the API passes the new evidence buffer to the internal
+API that is used to generate the self-signed certificate with the evidence embedded.
+Similar to the run-time claims support, the API appends the `runtime_custom_claims_buffer` to the
+`public_key` and passes the new buffer to the `oe_get_evidence` API.
+
+```diff
+oe_result_t oe_get_attestation_certificate_with_evidence_v3(
+    const oe_uuid_t* format_id,
+    const unsigned char* subject_name,
+    const uint8_t* private_key,
+    size_t private_key_size,
+    const uint8_t* public_key,
+    size_t public_key_size,
++   const uint8_t* runtime_custom_claims_buffer,
++   size_t runtime_custom_claims_buffer_size,
++   const uint8_t* inittime_custom_claims_buffer,
++   size_t inittime_custom_claims_buffer_size,
+    const void* optional_parameters,
+    size_t optional_parameters_size,
+    uint8_t** output_certificate,
+    size_t* output_certificate_size)
+{
+    oe_result_t result = OE_FAILURE;
+    uint8_t* evidence_buffer = NULL;
+    size_t evidence_buffer_size = 0;
++   uint8_t* custom_claims_buffer = NULL;
++   size_t custom_claims_buffer_size = 0;
++   uint8_t* evidence_with_inittime_custom_claims = NULL;
++   size_t evidence_with_inittime_custom_claims_size = 0;
+
+    OE_TRACE_VERBOSE("Calling oe_get_attestation_certificate_with_evidence_v3");
+    OE_TRACE_VERBOSE(
+        "generate evidence with hash from public_key_size=%d public_key key "
+        "=\n[%s]\n",
+        public_key_size,
+        public_key);
+
++   custom_claims_buffer_size = public_key_size + runtime_custom_claims_buffer_size;
++   custom_claims_buffer = oe_malloc(custom_claims_buffer_size);
++   if (!custom_claims_buffer)
++       OE_RAISE(OE_OUT_OF_MEMORY);
+
++   memcpy(custom_claims_buffer, public_key, public_key_size);
++   memcpy(
++       custom_claims_buffer + public_key_size,
++       runtime_custom_claims_buffer,
++       runtime_custom_claims_buffer_size);
+
+    result = oe_get_evidence(
+        format_id,
+        OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID,
+-       public_key,
+-       public_key_size,
++       custom_claims_buffer,
++       custom_claims_buffer_size,
+        optional_parameters,
+        optional_parameters_size,
+        &evidence_buffer,
+        &evidence_buffer_size,
+        NULL,
+        0);
+    OE_CHECK_MSG(
+        result, "oe_get_evidence failed with %s\n", oe_result_str(result));
+
++   evidence_with_inittime_custom_claims_size =
++       evidence_buffer_size + inittime_custom_claims_buffer_size;
++   evidence_with_inittime_custom_claims =
```

Comment:
Same question about the init time claims. How can we extract it without the size info?

Reply:
The design is based on the contract between `oe_get_attestation_certificate_with_evidence_v3` and `oe_verify_attestation_certificate_with_evidence_v3` APIs. Both APIs assume the evidence with the `OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID` format instead of raw report. The evidence with this format comes with a header that includes the size of the evidence.